### PR TITLE
Validations for data types used for machine types implemented + tests

### PR DESCRIPTION
**Pull request by hunterkepley**
This PR includes validations for each field of a machine type when parsing, to make sure they are the right datatype

This PR also includes a (passing) unit test for the validations

### CODE REVIEW

File: pkg/validations/validations.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+func validateDataType(resourceId, resourceField string, resourceValue interface{}, dataType reflect.Type) error {
+	if reflect.TypeOf(resourceValue) != dataType {
+		return errors.UserErrorf(errormsgs.SyntaxDataType, resourceId, resourceField)
+	}
+	return nil
+}
+
+func validateMachineTypeDataTypes(resource map[string]interface{}) error {
```
Comment (marcolan018) [inline pkg/validations/validations.go:19]: can we use constant for the field names?

Comment (hunterkepley) [reply pkg/validations/validations.go:19]: Good idea, i'll add that

File: pkg/convert/convert.go [hunk 2]
```diff
@@ -66,51 +72,17 @@ func ResourcesToYamlMachineTypes(resources []map[string]interface{}, outputFile
...
-				resource["cloud_provider_id"].(string) + "\n      cpu_cores: " +
-				resource["cpu_cores"].(string) + "\n      memory: " + resource["memory"].(string) +
+				"\n      name: " + resource[Name].(string) + "\n      cloud_provider_id: " +
+				resource[CloudProviderId].(string) + "\n      cpu_cores: " +
+				resource[CpuCores].(string) + "\n      memory: " + resource[Memory].(string) +
 				"\n      category: " +
-				resource["category"].(string) + "\n      size: " + resource["size_pretty"].(string) +
+				resource[Category].(string) + "\n      size: " + resource[SizePretty].(string) +
 				"\n      generic_name: " +
-				resource["generic_name"].(string) + "\n")
```
Comment (ziccardi) [inline pkg/convert/convert.go]: I think that formatting this code differently would improve readability
```go
			machineTypeString := []byte("    - id: " + id +
				"\n      name: " + resource[Name].(string) +
				"\n      cloud_provider_id: " + resource[CloudProviderId].(string) +
				"\n      cpu_cores: " + resource[CpuCores].(string) +
				"\n      memory: " + resource[Memory].(string) +
				"\n      category: " + resource[Category].(string) +
				"\n      size: " + resource[SizePretty].(string) +
				"\n      generic_name: " + resource[GenericName].(string) + "\n")
```

Another option: instead of string concatenation, you could use `fmt.Sprintf`
```go
			machineTypeString = []byte(fmt.Sprintf(
				"    - id: %s\n"+
					"      name: %s\n"+
					"      cloud_provider_id: %s\n"+
					"      cpu_cores: %s\n"+
					"      memory: %s\n"+
					"      category: %s\n"+
					"      size: %s\n"+
					"      generic_name: %s\n",
				id, resource[Name].(string), resource[CloudProviderId].(string), resource[CpuCores].(string),
				resource[Memory].(string), resource[Category].(string), resource[SizePretty].(string),
				resource[GenericName].(string)))
```

File: pkg/pkg_test.go [hunk 3]
```diff
@@ -22,9 +26,9 @@ var _ = Describe("Pkg", func() {
 			machineTypes, err = convert.CsvToResources("test/test.csv")
 			Expect(err).To(BeNil())
```
Comment (ziccardi) [inline pkg/pkg_test.go]: This code is not part of this PR, but it should be
```suggestion
			Expect(err).NotTo(HaveOccurred())
```

File: pkg/convert/convert.go [hunk 4]
```diff
@@ -25,21 +27,25 @@ func CsvToResources(resourcesFile string) ([]map[string]interface{}, error) {
...
+		resource[CpuCores] = record[4]
+		resource[Memory] = record[5]
+		resource[MemoryPretty] = record[6]
+		resource[Category] = record[7]
+		resource[CategoryPretty] = record[8]
+		resource[SizePretty] = record[9]
+		resource[ResourceType] = record[10]
+		resource[Active] = record[11]
 		finalResources = append(finalResources, resource)
 	}
```
Comment (ziccardi) [inline pkg/convert/convert.go]: I would prefer to write this differently, but feel free to ignore this comment if you prefer the current version

```go
	keys := []string{Id, GenericName, Name, CloudProviderId, CpuCores, Memory, MemoryPretty, Category, CategoryPretty, SizePretty, ResourceType, Active}

	// Parse resources from CSV
	for _, record := range records {
		resource := make(map[string]interface{})
		for i, key := range keys {
			resource[key] = record[i]
		}

		finalResources = append(finalResources, resource)
	}
```

**Comment by ziccardi** [comment]
Great job, thanks!
Can we add some tests to cover the `ResourcesToYamlMachineTypes`?

**Comment by hunterkepley** [comment]
thanks for reviewing @ziccardi -- Will add more tests for that func

File: pkg/pkg_test.go [hunk 5]
```diff
@@ -6,34 +6,115 @@ import (
...
+			Expect(machineType[SizePretty]).To(Equal("12xlarge"))
+			Expect(machineType[CpuCores]).To(Equal("48"))
 		})
 		It("Convert parsed resource into yaml", func() {
 			err := convert.ResourcesToYamlMachineTypes(machineTypes, outputFile)
-			Expect(err).To(BeNil())
+			Expect(err).ToNot(HaveOccurred())
 			testFile, err := os.ReadFile(outputFile)
 
 			Expect(testFile).To(ContainSubstring(machineTypeId))
```
Comment (ziccardi) [inline pkg/pkg_test.go]: `err` is assigned, but never tested.
I believe we should test that no error has occurred:
```suggestion
			testFile, err := os.ReadFile(outputFile)
			Expect(err).ToNot(HaveOccurred())
			Expect(testFile).To(ContainSubstring(machineTypeId))
```

File: pkg/pkg_test.go [hunk 6]
```diff
@@ -6,34 +6,115 @@ import (
...
+			err = convert.ResourcesToYamlMachineTypes(machineTypes, "otpt/test.configmap.yaml")
+			Expect(err).To(HaveOccurred())
+		})
+		It("Fail for each field supplied with the wrong data type", func() {
+			// Check function which is to be used by the parse command
+			machineTypes, err := convert.CsvToResources("test/badDataTypes.csv")
+			Expect(err).To(HaveOccurred())
+			Expect(machineTypes).To(BeNil())
+			// Check each specific field validation
+			machineTypes, err = convert.CsvToResources("test/test.csv")
```
Comment (ziccardi) [inline pkg/pkg_test.go:85]: I believe we should test for `err` to be occurred or not.

File: pkg/pkg_test.go [hunk 7]
```diff
@@ -6,34 +6,117 @@ import (
...
+				To(Equal(fakeDataTypeValidationError(machineTypeId, field).Error()))
+			field = CpuCores
+			Expect(makeFakeMachineTypeDataType(machineType, field, "aaa").Error()).
+				To(Equal(fakeDataTypeValidationError(machineTypeId, field).Error()))
+			field = CloudProviderId
+			Expect(makeFakeMachineTypeDataType(machineType, field, 1234).Error()).
+				To(Equal(fakeDataTypeValidationError(machineTypeId, field).Error()))
+			field = Name
+			Expect(makeFakeMachineTypeDataType(machineType, field, 1234).Error()).
+				To(Equal(fakeDataTypeValidationError(machineTypeId, field).Error()))
```
Comment (ziccardi) [inline pkg/pkg_test.go]: Why are we using the `field` variable instead of directly using the constant (`GenericName`, `SizePretty`, etc)?

Comment (hunterkepley) [reply pkg/pkg_test.go]: ah yeah, that is better, this was left over from before they were consts, will change

File: pkg/validations/validations.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	}
+	if err := validateDataType(resourceId, SizePretty, resource[SizePretty], reflect.TypeOf("")); err != nil {
+		return err
+	}
+	if err := validateDataType(resourceId, GenericName, resource[GenericName], reflect.TypeOf("")); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (ziccardi) [inline pkg/validations/validations.go:53]: I was going to suggest refactoring the code as below for DRYness and easier future updates. 
Looking at it, however, I'm under the impression readability would suffer. What do you think?

```suggestion
func validateMachineTypeDataTypes(resource map[string]interface{}) error {
	resourceId := resource[Id].(string)
	if err := validateDataType(resourceId, Name, resource[Name], reflect.TypeOf("")); err != nil {
		return err
	}
	if err := validateDataType(resourceId, CloudProviderId, resource[CloudProviderId],
		reflect.TypeOf("")); err != nil {
		return err
	}
	cores, err := strconv.Atoi(resource[CpuCores].(string))
	if err != nil {
		return errors.UserErrorf(errormsgs.SyntaxDataType, resourceId, CpuCores)
	}
	memory, err := strconv.ParseInt(resource[Memory].(string), 10, 64)
	if err != nil {
		return errors.UserErrorf(errormsgs.SyntaxDataType, resourceId, Memory)
	}

	mkValidation := func(resourceId, resourceField string, resourceValue interface{}, dataType reflect.Type) func() error {
		return func() error {
			return validateDataType(resourceId, resourceField, resourceValue, dataType)
		}
	}

	validations := []func() error{
		mkValidation(resourceId, CpuCores, cores, reflect.TypeOf(0)),
		mkValidation(resourceId, Memory, memory, reflect.TypeOf(int64(0))),
		mkValidation(resourceId, Category, resource[Category], reflect.TypeOf("")),
		mkValidation(resourceId, SizePretty, resource[SizePretty], reflect.TypeOf("")),
		mkValidation(resourceId, GenericName, resource[GenericName], reflect.TypeOf("")),
	}

	for _, validator := range validations {
		if err := validator(); err != nil {
			return err
		}
	}

	return nil
}
```

Comment (hunterkepley) [reply pkg/validations/validations.go:53]: hmm yeah, i think readability is just a little bit better with the original, this is a good suggestion though

**Review by ciaranRoche — APPROVED** [review]
